eth/block-validation/api.go

```diff
@@ -344,3 +346,89 @@ func (b *BlockAssemblerRequest) UnmarshalJSON(data []byte) error {
 
 	return nil
 }
+
+func (api *BlockValidationAPI) BlockAssembler(params *BlockAssemblerRequest) (*capella.ExecutionPayload, error) {
+	log.Info("BlockAssembler", "tobTxs", len(params.TobTxs.Transactions), "robPayload", params.RobPayload)
+	transactionBytes := make([][]byte, len(params.TobTxs.Transactions))
+	for i, txHexBytes := range params.TobTxs.Transactions {
+		transactionBytes[i] = txHexBytes[:]
+	}
+	decodedTobTxs, err := engine.DecodeTransactions(transactionBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	robBlock, err := engine.ExecutionPayloadV2ToBlock(params.RobPayload.ExecutionPayload)
+	if err != nil {
+		return nil, err
+	}
+
+	tobTxs := types.Transactions(decodedTobTxs)
+
+	// TODO - check for gas limits
+	// TODO - support for payouts
+
+	// TODO - if there are no TOB txs then we can just simulate the block rather then re-assembling it.
+	// These are TODOs for now to simplify integration testing
+
+	// check if there are any duplicate txs
+	// we can error out if there is a nonce gap
+	// TODO - don't error out, but drop the duplicate tx in the ROB block
+	seenTxMap := make(map[common.Hash]struct{})
+	for _, tx := range tobTxs {
+		// If we see nonce reuse in TOB then fail
+		if _, ok := seenTxMap[tx.Hash()]; ok {
+			return nil, errors.New("duplicate tx")
+		}
+		seenTxMap[tx.Hash()] = struct{}{}
+	}
+	for _, tx := range robBlock.Transactions() {
+		// if we see nonce re-use in TOB vs ROB then drop txs the txs in ROB
+		if _, ok := seenTxMap[tx.Hash()]; ok {
+			return nil, errors.New("duplicate tx")
+		}
+		seenTxMap[tx.Hash()] = struct{}{}
+	}
+
+	withdrawals := make(types.Withdrawals, len(params.RobPayload.ExecutionPayload.Withdrawals))
+	for i, withdrawal := range params.RobPayload.ExecutionPayload.Withdrawals {
```

Comment:
confused why this copy is needed ?

Reply:
to convert from `[]*Withdrawal` to `types.Withdrawal` type. Will check if I can avoid it